homeassistant/components/godice/__init__.py

```diff
@@ -0,0 +1,68 @@
+"""The godice integration."""
+
+from __future__ import annotations
+
+import logging
+
+import bleak
+import godice
+
+from homeassistant.components import bluetooth
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ADDRESS, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+
+from .const import CONF_SHELL, DATA_DICE, DATA_DISCONNECTED_BY_REQUEST_FLAG, DOMAIN
+
+PLATFORMS = [Platform.SENSOR]
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up godice from a config entry."""
+    _LOGGER.debug("Setup started")
+
+    def on_disconnect_callback(_ble_data):
+        _LOGGER.debug("on_disconnect_callback called")
+        is_disconnected_by_request = hass.data[DOMAIN][entry.entry_id][
+            DATA_DISCONNECTED_BY_REQUEST_FLAG
+        ]
+        if not is_disconnected_by_request:
+            hass.create_task(hass.config_entries.async_reload(entry.entry_id))
+
+    try:
+        ble_device = bluetooth.async_ble_device_from_address(
+            hass, entry.data[CONF_ADDRESS]
+        )
+        assert ble_device is not None
+        client = bleak.BleakClient(
+            ble_device, timeout=20, disconnected_callback=on_disconnect_callback
+        )
```

Comment:
This seems a bit too low level for an integration

Reply:
https://github.com/home-assistant/core/pull/113065#discussion_r1525645810 

Please check https://developers.home-assistant.io/docs/api_lib_index/.  While its mostly talks about HTTP requests, the same applies for BLE connection logic.

> One of the foundational rules of Home Assistant is that we do not include any protocol specific code. Instead, this code should be put into a standalone Python library and published to PyPI. This guide will describe how to get started with this!

Reply:
all device interaction is done via a lib https://pypi.org/project/godice
Here I only instantiate device connection with some HA-specific calls (`bluetooth.async_ble_device_from_address`).
I would not like to put `bluetooth.async_ble_device_from_address` into `godice` lib because the lib itself is portable and may work on any Python env instead of being specifically tied to HA

Reply:
Using the HA bluetooth APIs in the integration is completely fine and expected.  Its the bleak APIs that shouldn't be in the integration

Reply:
ok I agree. I'll need to push a change to the godice lib itself and come back here once that change is merged. Making this PR a draft meanwhile

Reply:
@bdraco I've made a fix, recheck please